source/common/http/conn_manager_impl.cc

```diff
@@ -1074,6 +1077,26 @@ void ConnectionManagerImpl::ActiveStreamDecoderFilter::
   parent_.connection_manager_.stats_.named_.downstream_flow_control_resumed_reading_total_.inc();
 }
 
+void ConnectionManagerImpl::ActiveStreamDecoderFilter::addDownstreamWatermarkCallbacks(
+    DownstreamWatermarkCallbacks& watermark_callbacks) {
+  // This is called exactly once per stream, by the router filter.
+  // If there's ever a need for another filter to subscribe to watermark callbacks this can be
+  // turned into a vector.
+  ASSERT(parent_.watermark_callbacks_ == nullptr);
+  parent_.watermark_callbacks_ = &watermark_callbacks;
+  for (uint32_t i = 0; i < parent_.high_watermark_count_; ++i) {
```

Comment:
Sorry, you are right. I read this wrong the first time (for some reason I though we were iterating through all the streams here, never mind!). I might drop in a comment specific to the iteration talking about how we need to indicate above high watermark 0 or more times.